Implement insert overwrite unpartitioned table

xumingming opened the pull request:
For #510, Implement insert overwrite unpartitioned table

martint commented:
@cla-bot check

xumingming commented:
Any chance this PR get reviewed?

electrum commented:
@xumingming My sincere apologies for the delay. I am reviewing this now.

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveWriterFactory.java:
```diff
@@ -95,6 +96,7 @@
 
 public class HiveWriterFactory
 {
+    private static final io.airlift.log.Logger LOG = io.airlift.log.Logger.get(HiveWriterFactory.class);
```
This appears not to be used

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveWriterFactory.java:
```diff
@@ -345,8 +347,23 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
+                        writeInfo = locationService.getTableWriteInfo(locationHandle, true);
+                    }
+                    else if (insertExistingPartitionsBehavior == InsertExistingPartitionsBehavior.ERROR) {
+                        throw new PrestoException(HIVE_TABLE_READ_ONLY,
```
Let's make this error message identical to the above
```java
throw new PrestoException(HIVE_TABLE_READ_ONLY, "Unpartitioned Hive tables are immutable");
```

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveWriterFactory.java:
```diff
@@ -345,8 +347,23 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
+                                "Cannot insert into an readonly unpartitioned Hive table.");
+                    }
+                    else {
+                        throw new IllegalArgumentException(format("Unsupported insert existing table behavior: %s",
```
Nit: don't wrap here. Also, if there is only one argument at the end, use string concatenation:
```java
throw new IllegalArgumentException("Unsupported insert existing table behavior: " + insertExistingPartitionsBehavior);

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveMetadata.java:
```diff
@@ -85,4 +101,52 @@ public void testCreateMixedPredicate()
     }
+
+    @Test
+    public void testGetTablePrincipalPrivileges()
```
What we want to test here is the conversion from `Set<HivePrivilegeInfo>` in `getTablePrincipalPrivileges()`. This is difficult because it mixes the conversion with fetching partitions from the metastore.

Instead, we can move the conversion elsewhere, such as helper method on `PrincipalPrivileges`:
```java
public static PrincipalPrivileges fromHivePrivilegeInfo(Set<HivePrivilegeInfo> privileges)
```
Then the method is easy to test without needing to mock the metastore.

dain replied:
I think you can use the `FileHiveMetastore` with the `TestingHdfsEnvironment ` for this test.

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveLocationService.java:
```diff
@@ -0,0 +1,119 @@
+
+        public Assertion(LocationHandle locationHandle, boolean overwrite)
+        {
+            HdfsEnvironment hdfsEnvironment = mock(HdfsEnvironment.class);
```
Rather than `mock()`, reuse `TestingHdfsEnvironment` in `TestBackgroundHiveSplitLoader` (make it public)

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveLocationService.java:
```diff
@@ -0,0 +1,119 @@
+public class TestHiveLocationService
+{
+    @Test
+    public void testGetTableWriteInfo_append()
```
Nit: use an underscore violates the standard Java method naming convention. Name this `testGetTableWriteInfoAppend()`. I do agree it's more readable in this case, but it looks out of place in our code base.

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveLocationService.java:
```diff
@@ -0,0 +1,119 @@
+    @Test
+    public void testGetTableWriteInfo_append()
+    {
+        assertThat(locationHandle(LocationHandle.WriteMode.STAGE_AND_MOVE_TO_TARGET_DIRECTORY), false)
```
Nit: Static import all the `WriteMode` constants to make the tests more readable

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveLocationService.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private static LocationHandle locationHandle(LocationHandle.WriteMode writeMode,
+            String targetPath, String writePath)
```
Nit: don't wrap the parameters here (or put each one on a separate line)

dain commented on pom.xml:
```diff
@@ -1126,6 +1126,20 @@
             </dependency>
+
+            <dependency>
+                <groupId>org.mockito</groupId>
```
We have found over the years that mocking frameworks lead to brittle tests that really don't end up testing much.  I know there are good ways to use mocking frameworks, but it is too easy to use them in ways that ends up with difficult to maintain tests.  As a result, we have actively avoided them in Presto.  

Instead of mocking frameworks, we typically have a local, in-memory, version of interfaces or services for testing.  For example, TestingAccessControlManager or TestingMetadata.  I'll comment below on how I think we can test this without introducing mockito to Presto.

dain commented on presto-hive/src/test/java/io/prestosql/plugin/hive/metastore/TestSemiTransactionalHiveMetastore.java:
```diff
@@ -0,0 +1,368 @@
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestSemiTransactionalHiveMetastore
```
I think you can use the `FileHiveMetastore` with the `TestingHdfsEnvironment ` for this test.

xumingming commented:
Thanks @electrum @dain for the reviewing, I'll update the code in the recent days.

xumingming commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1347,18 +1352,37 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
         for (PartitionUpdate partitionUpdate : partitionUpdates) {
             if (partitionUpdate.getName().isEmpty()) {
                 // insert into unpartitioned table
+                if (!table.get().getStorage().getStorageFormat().getInputFormat().equals(
```
Will do.

I'd like to discuss it a little bit here, I wrap because the line is too long to read, and I find that the presto code are often very very long in one line, it does not annoy you guys? Or because you are all using very big monitor?

xumingming commented:
Hi @electrum , @dain , I have cleaned up the codes:

1. Removed mockito
2. Cleaned up code style.
3. Removed two tests -- Without Mockito, it requires a fully functional TestingHdfsEnviroment & TestingFileSystem.

dain commented on pom.xml:
```diff
@@ -1126,6 +1126,13 @@
             </dependency>
+
+            <dependency>
+                <groupId>org.objenesis</groupId>
```
This doesn't seem to be used anywhere.

electrum commented:
I found some issues as I was testing this before merging, so I fixed them up and added a few more tests. This is now merged as #924

Thanks for the contribution!
